zkevm-circuits/src/evm_circuit/op_execution/push.rs

```diff
@@ -0,0 +1,307 @@
+use super::super::{
+    BusMappingLookup, Case, Cell, Constraint, CoreStateInstance, ExecutionStep,
+    FixedLookup, Lookup, Word,
+};
+use super::{CaseAllocation, CaseConfig, OpExecutionState, OpGadget};
+use halo2::plonk::Error;
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Expression};
+use std::{array, convert::TryInto};
+
+#[derive(Clone, Debug)]
+struct PushSuccessAllocation<F> {
+    case_selector: Cell<F>,
+    word: Word<F>,
+    selectors: [Cell<F>; 32], /* whether its PUSH1, ..., or PUSH32 ([1, 1,
+                               * 0, ..., 0] means PUSH2) */
+}
+
+#[derive(Clone, Debug)]
+pub struct PushGadget<F> {
+    success: PushSuccessAllocation<F>,
+    stack_overflow: Cell<F>, // case selector
+    out_of_gas: (
+        Cell<F>, // case selector
+        Cell<F>, // gas available
+    ),
+}
+
+impl<F: FieldExt> OpGadget<F> for PushGadget<F> {
+    const RESPONSIBLE_OPCODES: &'static [u8] = &[
+        96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110,
+        111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124,
+        125, 126, 127,
+    ];
+
+    const CASE_CONFIGS: &'static [CaseConfig] = &[
+        CaseConfig {
+            case: Case::Success,
+            num_word: 1,
+            num_cell: 32, // for PUSH selectors
+            will_resume: false,
+        },
+        CaseConfig {
+            case: Case::StackOverflow,
+            num_word: 0,
+            num_cell: 0,
+            will_resume: true,
+        },
+        CaseConfig {
+            case: Case::OutOfGas,
+            num_word: 0,
+            num_cell: 0,
+            will_resume: true,
+        },
+    ];
```

Comment:
This kinda reforces my previous argument that it doesn't seem to fit really well to have CaseConfigs as a const associated to the gadget since it can change depending on the instance.